### ISSUE TITLE
Projection can contain out of scope variables

### ISSUE THREAD

**Issue by joachimvh** [joachimvh/SPARQLAlgebra.js#89]
Running the following query through the library:
```sparql
PREFIX : <http://example.com/>
SELECT * WHERE {
        ?x :p ?n
        FILTER NOT EXISTS {
                ?x :q ?m .
                FILTER(?n = ?m)
        }
}
```
and then back to SPARQL, results in:
```sparql
SELECT ?x ?n ?m WHERE {
  ?x <http://example.com/p> ?n.
  FILTER(NOT EXISTS {
    ?x <http://example.com/q> ?m.
    FILTER(?n = ?m)
  })
}
```
`?m` should not be there according to the example: https://www.w3.org/TR/sparql11-query/#idp899488

The function that calculates the in scope variables is probably too flexible and should follow the following table more closely: https://www.w3.org/TR/sparql11-query/#variableScope

I see that the function that translates blank nodes also depends on the output of that so will have to make sure that doesn't break anything.

**Comment by rubensworks** [comment]
Heh, I just got a report that may be caused by this: https://gitter.im/comunica/Lobby?at=61eac0359a3354540633e751

**Comment by rubensworks** [comment]
Thought it might be related to https://github.com/comunica/comunica/issues/866, but probably not.

**Comment by joachimvh** [comment]
> Thought it might be related to [comunica/comunica#866](https://github.com/comunica/comunica/issues/866), but probably not.

I agree, in that issue the correct variables get projected but Comunica doesn't create them in the bindings for some reason.

**Comment by joachimvh** [comment]
Fixed in b44a44003af28f4e3ff1f031c5a50eda3fde6a39 . As a bonus projected variables will now be sorted alphabetically if no order was defined (SELECT * was used).

Will do a new patch release.
